### DISCUSSION TITLE
Client traffic not going through a gateway

### DISCUSSION THREAD

**Discussion by seldfv** [strongswan/strongswan#1293] [Q&A]
Good afternoon,

I've been having a lot of difficulties with routing traffic using strongswan.
The client can successfully connect to the VPN, but doesn't know where to go from there.

172.16.10.0 ----- | 172.16.10.111 - 178.X.X.X | === | x.x.x.x | -- 172.16.10.152
--- subnet ----------------  ens3   ----------- ens4 ---------------client----client virtual IP
----------------------------strongswan machine

I've made a post about something similar where clients couldn't access the local LAN behind the VPN : https://github.com/strongswan/strongswan/discussions/1124
Right after that I went to vacations and did other projects, so I still have this problem and I tried to pinpoint the issue.

What I found after trying some new things today : 

After doing a tracert on my Windows client, the traffic get immediately lost. So I went an added (a bunch) of routes : 

```
IPv4 Table de routage
===========================================================================
Itinéraires actifs :
Destination réseau    Masque réseau  Adr. passerelle   Adr. interface Métrique
          0.0.0.0          0.0.0.0      172.20.10.1      172.20.10.2     50
        127.0.0.0        255.0.0.0         On-link         127.0.0.1    331
        127.0.0.1  255.255.255.255         On-link         127.0.0.1    331
  127.255.255.255  255.255.255.255         On-link         127.0.0.1    331
       172.16.0.0      255.255.0.0         On-link     172.16.10.152     36
      172.16.10.0    255.255.255.0         On-link     172.16.10.152     36
      172.16.10.0    255.255.255.0    178.X.X.X      172.20.10.2     51
    172.16.10.152  255.255.255.255         On-link     172.16.10.152    291
    172.16.10.255  255.255.255.255         On-link     172.16.10.152    291
   172.16.255.255  255.255.255.255         On-link     172.16.10.152    291
      172.20.10.0  255.255.255.240         On-link       172.20.10.2    306
      172.20.10.2  255.255.255.255         On-link       172.20.10.2    306
     172.20.10.15  255.255.255.255         On-link       172.20.10.2    306
      178.X.X.0  255.255.255.224         On-link       172.20.10.2     51
     178.X.X.Z  255.255.255.255         On-link       172.20.10.2    306
    178.X.X.X  255.255.255.255      172.20.10.1      172.20.10.2     51
      192.168.1.0    255.255.255.0         On-link     172.16.10.152     36
    192.168.1.255  255.255.255.255         On-link     172.16.10.152    291
      198.168.0.0      255.255.0.0         On-link     172.16.10.152     36
  198.168.255.255  255.255.255.255         On-link     172.16.10.152    291
        224.0.0.0        240.0.0.0         On-link         127.0.0.1    331
        224.0.0.0        240.0.0.0         On-link       172.20.10.2    306
        224.0.0.0        240.0.0.0         On-link     172.16.10.152    291
  255.255.255.255  255.255.255.255         On-link         127.0.0.1    331
  255.255.255.255  255.255.255.255         On-link       172.20.10.2    306
  255.255.255.255  255.255.255.255         On-link     172.16.10.152    291
```

172.20.10.2 is the WiFi ip address of my 4g shared connection.

ipconfig /all : 

```
Carte PPP Linkso VPN :

   Suffixe DNS propre à la connexion. . . :
   Description. . . . . . . . . . . . . . : Linkso VPN
   Adresse physique . . . . . . . . . . . :
   DHCP activé. . . . . . . . . . . . . . : Non
   Configuration automatique activée. . . : Oui
   Adresse IPv4. . . . . . . . . . . . . .: 172.16.10.152(préféré)
   Masque de sous-réseau. . . . . . . . . : 255.255.255.255
   Passerelle par défaut. . . . . . . . . :
   Serveurs DNS. . .  . . . . . . . . . . : 172.16.10.72
                                       172.16.10.73
   NetBIOS sur Tcpip. . . . . . . . . . . : Activé
```

What wireshark shows me : 

![wireshark](https://user-images.githubusercontent.com/113189164/189682861-369d643a-5231-438a-b0b4-3317ea2b92d9.png)

swanctl.conf : 

```
connections {

eap-17216800 {

        local_addrs = 178.X.X.X
        pools = radius
        version = 2
        send_cert = always
        proposals = aes256-sha384-prfsha384-ecp384

                local {
                auth = rsa
                certs = server-cert.pem
                id = 172.16.80.241
                }

                remote {
                        auth = eap-radius
                        eap_id = %any
                        id = %any
                        groups = 17216800
                }

                children {
                        net {
                                local_ts = 172.16.10.0/24
                                esp_proposals = aes256gcm16-ecp384,aes256
                        }
                }
        }
}
secrets {

        rsa {
        file = server-key.pem
        }
}
```

strongswan.conf : 

```
swanctl {
  load = pem rsa pkcs1 x509 revocation constraints pubkey openssl random
}


charon-systemd {
        load = farp random nonce aes eap-mschapv2 openssl sha1 sha2 md5 pem pkcs1 curve25519 gmp x509 curl revocation hmac kdf vici kernel-netlink socket-default eap-identity eap-radius

        plugins {
                eap-radius {
                        class_group = yes
                        secret = ################
                        server = 127.0.0.1
                }
        }
}
```

The client get it's IP address from Radius : 

/etc/raddb/users : 

```
test1.vpn       User-Name == test1.vpn
        Class = "17216100",
        Framed-IP-Address += 172.16.10.152,
        MS-Primary-DNS-Server += 172.16.10.72,
        MS-Primary-DNS-Server += 172.16.10.73,
        Fall-Through = Yes
```

My firewalld : 

public (active)
  target: default
  icmp-block-inversion: no
  interfaces: ens4
  sources: 
  services: cockpit dhcpv6-client ipsec ssh
  ports: 500/udp 4500/udp
  protocols: 
  forward: no
  masquerade: yes
  forward-ports: 
  source-ports: 
  icmp-blocks: 
  rich rules: 
	rule protocol value="esp" accept

internal (active)
  target: default
  icmp-block-inversion: no
  interfaces: ens3
  sources: 
  services: cockpit dhcpv6-client mdns samba-client ssh
  ports: 
  protocols: 
  forward: no
  masquerade: yes
  forward-ports: 
  source-ports: 
  icmp-blocks: 
  rich rules: 

Edit : sysctl -p
[root@--------------- ~]# sysctl -p
net.ipv4.ip_forward = 1
net.ipv4.ip_no_pmtu_disc = 1
net.ipv4.conf.all.accept_redirects = 0
net.ipv4.conf.all.send_redirects = 0

i've already tried https://docs.strongswan.org/docs/5.9/howtos/forwarding.html recommendations by doing NAT rules in the firewall (maybe I am missing a direct rule that NAT the virtual ip from the 172.16.10.0/24 subnet to the ens4 interface... I've read a lot of RHEL installation but nothing much comes out with firewalld).

Thanks in advance for any help, that would be really appreciated.
Best regards,

**Comment by seldfv** [comment]
I've stopped using firewalld for iptables but the problem is persistent. My iptables rules : 

```
*filter
:INPUT ACCEPT [192:16766]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [140:17329]
-A INPUT -p tcp -m multiport --dports 22 -j f2b-sshd
-A INPUT -i ens4 -p esp -j ACCEPT
-A INPUT -i ens4 -p ah -j ACCEPT
-A INPUT -i ens4 -p udp -m udp --sport 500 --dport 500 -j ACCEPT
-A INPUT -i ens4 -p udp -m udp --sport 4500 --dport 4500 -j ACCEPT
-A FORWARD -s 172.16.10.0/32 -d 172.16.10.0/24 -i ens4 -m policy --dir in --pol ipsec --reqid 2 --proto esp -j ACCEPT
-A FORWARD -s 172.16.10.0/24 -d 172.16.10.0/32 -o ens4 -m policy --dir out --pol ipsec --reqid 2 --proto esp -j ACCEPT
-A OUTPUT -o ens4 -p esp -j ACCEPT
-A OUTPUT -o ens4 -p ah -j ACCEPT
-A OUTPUT -o ens4 -p udp -m udp --sport 500 --dport 500 -j ACCEPT
-A OUTPUT -o ens4 -p udp -m udp --sport 4500 --dport 4500 -j ACCEPT
COMMIT
# Completed on Tue Sep 13 09:13:35 2022
# Generated by iptables-save v1.8.4 on Tue Sep 13 09:13:35 2022
*security
:INPUT ACCEPT [252:21853]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [219:65798]
COMMIT
# Completed on Tue Sep 13 09:13:35 2022
# Generated by iptables-save v1.8.4 on Tue Sep 13 09:13:35 2022
*raw
:PREROUTING ACCEPT [335:25265]
:OUTPUT ACCEPT [223:66870]
COMMIT
# Completed on Tue Sep 13 09:13:35 2022
# Generated by iptables-save v1.8.4 on Tue Sep 13 09:13:35 2022
*mangle
:PREROUTING ACCEPT [337:25369]
:INPUT ACCEPT [257:22113]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [228:67534]
:POSTROUTING ACCEPT [228:67534]
COMMIT
# Completed on Tue Sep 13 09:13:35 2022
# Generated by iptables-save v1.8.4 on Tue Sep 13 09:13:35 2022
*nat
:PREROUTING ACCEPT [103:5001]
:INPUT ACCEPT [23:1745]
:POSTROUTING ACCEPT [18:1409]
:OUTPUT ACCEPT [18:1409]
-A POSTROUTING -m policy --dir out --pol ipsec -j ACCEPT
-A POSTROUTING -s 172.16.10.0/24 -o ens3 -m policy --dir out --pol ipsec -j ACCEPT
-A POSTROUTING -s 172.16.10.0/24 -o ens3 -j MASQUERADE
COMMIT
# Completed on Tue Sep 13 09:13:35 2022
```

**Comment by seldfv** [comment]
Update : I tried it on my android phone and it worked, so it's definitely something with the Windows client

**Reply by tobiasbrunner** [reply]
Windows clients do some weird stuff when using split tunneling (to some degree depends on the negotiated traffic selectors and virtual IPs, some Windows options, possibly manually configured routes etc.). The docs have [some information](https://docs.strongswan.org/docs/5.9/howtos/forwarding.html#_split_tunneling) (also some [here](https://docs.strongswan.org/docs/5.9/interop/windowsClients.html#_split_routing_since_windows_10)), but it's been a while since anybody did a proper analysis of how current Windows versions behave in specific setups.

**Reply by seldfv** [reply]
The issue was indeed coming from the Windows client.

I had to have an outgoing firewall rule that allow all services and only the protocol 4 that is IPv4 encap  and add the public IP address 178.X.X.X as the endpoint (edit : only for the private profile). I finally get result listening to the public interface : 

**15:46:40.428641 IP 10.10.10.2 > XXXXX.XXXXXX.local: ICMP echo request, id 1, seq 305, length 40**

I changed the client ip address 172.16.10.152 to something more casual that is 10.10.10.2

Now I'd like to ping the client via the strongswan server : 

**traceroute to 10.10.10.2 (10.10.10.2), 30 hops max, 60 byte packets
 1  _gateway (178.X.X.193)  2.087 ms  1.946 ms  1.792 ms**

And it gets lost after this. My thinking is that I have to add a route on the internal gateway (172.16.10.254) for the gateway to let him know that I have to go through him to get to the client. It will take some time to ask the admin to add the correct route.


I feel like I just did the easy solution. Hope it will help someone !
